### PR TITLE
implement KustomizeProvider

### PR DESCRIPTION
**Pull request by Liujingfang1**

### CODE REVIEW

**Comment by Liujingfang1** [comment]
@pwittrock I found a similar function from wirek8s package https://github.com/kubernetes-sigs/cli-experimental/blob/35eeb51394dac4e6ce40c523b6081dc95600c303/internal/pkg/wirecli/wirek8s/wirek8s.go#L87. Is this function intended to be replaced by function from internal/pkg/resourceconfig?

File: internal/pkg/apply/apply.go [hunk 1]
```diff
@@ -39,13 +42,42 @@ type Result struct {
 // Do executes the apply
 func (a *Apply) Do() (Result, error) {
 	fmt.Fprintf(a.Out, "Doing `cli-experimental apply`\n")
-	pods, err := a.Clientset.CoreV1().Pods("default").List(metav1.ListOptions{})
+	cl, err := client.New(config.GetConfigOrDie(), client.Options{})
```
Comment (pwittrock) [inline internal/pkg/apply/apply.go]: This should be initialized by the injector and injected

File: internal/pkg/apply/apply.go [hunk 2]
```diff
@@ -39,13 +42,42 @@ type Result struct {
...
+			fmt.Fprintf(a.Out, "Not able to convert runtime object to unstructured\n")
+			continue
+		}
+		live := &unstructured.Unstructured{}
+		live.SetGroupVersionKind(u.GroupVersionKind())
+		err := cl.Get(context.TODO(), client.ObjectKey{
+			Namespace: u.GetNamespace(),
+			Name:      u.GetName(),
+		}, live)
+		if err == nil {
```
Comment (pwittrock) [inline internal/pkg/apply/apply.go]: This is just a placeholder for development until the new Apply command is wired in?

File: internal/pkg/apply/apply.go [hunk 3]
```diff
@@ -39,13 +42,42 @@ type Result struct {
...
-		fmt.Fprintf(a.Out, "Pod %s\n", p.Name)
+	for _, resource := range a.Resources {
+		u, ok := resource.(*unstructured.Unstructured)
+		if !ok {
+			fmt.Fprintf(a.Out, "Not able to convert runtime object to unstructured\n")
+			continue
+		}
+		live := &unstructured.Unstructured{}
+		live.SetGroupVersionKind(u.GroupVersionKind())
+		err := cl.Get(context.TODO(), client.ObjectKey{
```
Comment (pwittrock) [inline internal/pkg/apply/apply.go]: Why not use the dynamic client?  It is pretty good for unstructured objects.

File: internal/pkg/apply/apply.go [hunk 4]
```diff
@@ -39,13 +42,42 @@ type Result struct {
...
+			continue
+		}
+		fmt.Fprintf(a.Out, "failed to get live object\n%v\n", err)
 	}
 
 	return Result{Resources: a.Resources}, nil
 }
+
+func isNotFoundError(e error) bool {
+	return strings.Contains(e.Error(), "not found") || strings.Contains(e.Error(), "not find")
```
Comment (pwittrock) [inline internal/pkg/apply/apply.go]: Can we check the error type to see if it is a not found error?

File: internal/pkg/apply/apply.go [hunk 5]
```diff
@@ -39,13 +42,42 @@ type Result struct {
...
+			continue
+		}
+		if isNotFoundError(err) {
+			err = cl.Create(context.Background(), resource)
+			if err != nil {
+				fmt.Fprintf(a.Out, "failed to create object:\n%v\n", err)
+			}
+			continue
+		}
+		fmt.Fprintf(a.Out, "failed to get live object\n%v\n", err)
```
Comment (pwittrock) [inline internal/pkg/apply/apply.go]: Why are we printing this?

File: internal/pkg/apply/apply.go [hunk 6]
```diff
@@ -39,13 +42,42 @@ type Result struct {
...
+			err = cl.Update(context.Background(), resource)
+			if err != nil {
+				fmt.Fprintf(a.Out, "failed to update object:\n%v\n", err)
+			}
+			continue
+		}
+		if isNotFoundError(err) {
+			err = cl.Create(context.Background(), resource)
+			if err != nil {
+				fmt.Fprintf(a.Out, "failed to create object:\n%v\n", err)
```
Comment (pwittrock) [inline internal/pkg/apply/apply.go]: This should go to STDERR

File: internal/pkg/apply/apply.go [hunk 7]
```diff
@@ -39,13 +42,42 @@ type Result struct {
...
+		live := &unstructured.Unstructured{}
+		live.SetGroupVersionKind(u.GroupVersionKind())
+		err := cl.Get(context.TODO(), client.ObjectKey{
+			Namespace: u.GetNamespace(),
+			Name:      u.GetName(),
+		}, live)
+		if err == nil {
+			err = cl.Update(context.Background(), resource)
+			if err != nil {
+				fmt.Fprintf(a.Out, "failed to update object:\n%v\n", err)
```
Comment (pwittrock) [inline internal/pkg/apply/apply.go]: SHould go to stderr

File: internal/pkg/apply/apply.go [hunk 8]
```diff
@@ -39,13 +42,42 @@ type Result struct {
...
 	if err != nil {
-		return Result{}, err
+		fmt.Fprintf(a.Out, "failed to create a client\n")
 	}
-	for _, p := range pods.Items {
-		fmt.Fprintf(a.Out, "Pod %s\n", p.Name)
+	for _, resource := range a.Resources {
+		u, ok := resource.(*unstructured.Unstructured)
+		if !ok {
+			fmt.Fprintf(a.Out, "Not able to convert runtime object to unstructured\n")
```
Comment (pwittrock) [inline internal/pkg/apply/apply.go]: Maybe we should just change the type of Resources to be a list of Unstructured pointers.

File: internal/pkg/wirecli/wirek8s/wirek8s.go [hunk 9]
```diff
@@ -74,35 +73,19 @@ func NewRestConfig(master clik8s.MasterURL, path clik8s.KubeConfigPath) (*rest.C
...
-	// Kustomization file.  Don't allow recursing on directories with raw Resource Config,
-	// should use a kustomization.yaml instead.
-	if fi.IsDir() {
-		k, err := doDir(p, sysFs)
-		if err != nil {
-			return nil, err
-		}
-		values = append(values, k...)
-		return values, nil
+	kp := resourceconfig.NewKustomizeProvider()
```
Comment (pwittrock) [inline internal/pkg/wirecli/wirek8s/wirek8s.go]: This should be passed in to the function rather than initialized.

**Comment by Liujingfang1** [comment]
Restrict this PR to focus on the KustomizeProvider. I'll update the apply code using dynamic client in a different PR.

**Pull request closed without merging**